### ISSUE TITLE
Support for Metal on macOS/iOS

### ISSUE THREAD

**Issue by gh0sti** [mupen64plus/mupen64plus-core#574]
Dear Team,
With macOS 10.14 Apple has decided to deprecate OpenGL/CL, in the near future mupen64 will not be able to run on macOS or iOS
![image](https://user-images.githubusercontent.com/4202657/40941139-9bce10e2-680f-11e8-93b8-c99c131ea0ec.png)

Will you guys look into possibly porting mupen64 to metal?

**Comment by loganmc10** [comment]
The GFX plugin would need to support Metal, currently no n64 GFX plugin supports Metal, so there is no point in mupen64plus supporting that API. You need to find a GFX plugin developer willing to add that to a GFX plugin

**Comment by rlabrecque** [comment]
Is there any vulkan based plugins? Could probably use MoltenVK?

**Comment by loganmc10** [comment]
No there is not currently a Vulkan based GFX plugin either (retroarch has one but it is not maintained)

**Comment by dankcushions** [comment]
deprecated!=removed. it's just not going to be supported/updated. arguably it's been that way for quite a while on osx/ios.

**Comment by richard42** [comment]
I hate to say it, but Apple is the new Microsoft. This was a very bad thing for them to do to the software development community. It's one thing for them to deprecate their own obsolete technologies (32-bit quicktime), but to abandon open standards and force everyone to re-write all their code, is another.

**Comment by Narann** [comment]
> It's one thing for them to deprecate their own obsolete technologies (32-bit quicktime), but to abandon open standards and force everyone to re-write all their code, is another.

I totally agree with this. I'm not in favor of proprietary API because you have no idea how it will evolve and you can have invest a lot of testing and time to an API to see it finally put in a trash. But Microsoft did a very decent job with D3D9 (letting peoples switch to other API OGL/VK/D3D11+) . But Apple is saying *f.ck you* to peoples writing and maintaining softwares. That's crazy how a short term view it can be.

What I'm afraid of would be the monkey patch coming in PR saying: "Metal support, please merge or this mean you have no respect for your users...".

Maybe a polite "That's cool, please fork it." can be an answer, or we would have code to remove in few years.

I don't know how other emu devs will handle this. It require resources. Dolphin devs would certainly add support (then maybe drop it because no one want to maintain it, [like they did with D3D12](https://github.com/dolphin-emu/dolphin/pull/4424)). Maybe libretro is already doing the job. I don't know if GlideN64 have any plan.

**Comment by mudlord** [comment]
Well GLideN64 should have a plan. They got themselves into this mess.

**Comment by JoaoPinheiro** [comment]
What are the pieces that would need to be rewritten for Metal if someone would like to pick up the task?

**Comment by richard42** [comment]
Well the graphics handling software is split in a few different places in Mupen64plus. A relatively small amount of code is in the core library (called the Video Extension) and optionally in the front-end application as well, to create and manage graphics contexts. This is necessary in order to support features like embedded OpenGL windows in a GUI app. It would be nice to extend this code to support the creation and management of Metal contexts as well, though it's probably not necessary. The bigger part of the code is in the video plugin. A skilled developer could either A) start with an existing plugin written for a different graphics API, and adapt it to Metal, or B) write a new one from scratch. These would both take a good amount of work from a very talented developer, but option A would probably take much less work than B. If a developer started with a video plugin that already used shaders for everything, then it might not be terribly hard.
